x/claim/keeper/claim.go

```diff
@@ -0,0 +1,161 @@
+	defer iterator.Close()
+
+	balances := []banktypes.Balance{}
+	for ; iterator.Valid(); iterator.Next() {
+		coins, err := sdk.ParseCoinsNormalized(string(iterator.Value()))
+		if err != nil {
+			panic(err)
+		}
+		addrBz := iterator.Key()[len(types.ClaimableStoreKey):]
+		addr := sdk.AccAddress(addrBz)
+		balances = append(balances, banktypes.Balance{
+			Address: addr.String(),
+			Coins:   coins,
+		})
+	}
+	return balances
+}
+
+// GetClaimable returns claimable amount for an address
+func (k Keeper) GetClaimable(ctx sdk.Context, addr string) (sdk.Coins, error) {
+	store := ctx.KVStore(k.storeKey)
+	prefixStore := prefix.NewStore(store, []byte(types.ClaimableStoreKey))
+	if !prefixStore.Has([]byte(addr)) {
+		return sdk.Coins{}, nil
+	}
+	bz := prefixStore.Get([]byte(addr))
+	coins, err := sdk.ParseCoinsNormalized(string(bz))
+	if err != nil {
+		return coins, err
+	}
+
+	params, err := k.GetParams(ctx)
+	if err != nil {
+		return coins, err
+	}
+
+	goneTime := ctx.BlockTime().Sub(params.AirdropStart)
+	if goneTime <= params.DurationUntilDecay {
+		// still not the time for decay
+		return coins, nil
+	}
+
+	if goneTime > params.DurationUntilDecay+params.DurationOfDecay {
+		// airdrop time passed
+		return sdk.Coins{}, nil
+	}
+
+	claimableCoins := sdk.Coins{}
+	monthlyDecayPercent := 10
+	monthDuration := time.Hour * 24 * 30
+	// Positive, since goneTime > params.DurationUntilDecay
+	decayTime := goneTime - params.DurationUntilDecay
+	for _, coin := range coins {
+		// decayPercent = decay_percent_per_month * (time_decayed) / (1 month)
+		// claimable_percent = 100 - decayPercent
+		// claimable_amt = claimable_coins * claimable_percent / 100
+		decayPercent := monthlyDecayPercent * int(decayTime) / int(monthDuration)
+		claimablePercent := int64(100) - int64(decayPercent)
+		claimableAmt := big.NewInt(0).Div(coin.Amount.Mul(sdk.NewInt(claimablePercent)).BigInt(), big.NewInt(100))
+		claimableCoinsWithDecay := sdk.NewCoin(coin.Denom, sdk.NewIntFromBigInt(claimableAmt))
+		claimableCoins = sdk.Coins{claimableCoinsWithDecay}
+	}
+
+	return claimableCoins, nil
+}
+
+// GetClaimablesByActivity returns the withdrawal amount from users' airdrop amount and activity made
+func (k Keeper) GetWithdrawableByActivity(ctx sdk.Context, addr string) (sdk.Coins, error) {
+	address, err := sdk.AccAddressFromBech32(addr)
+	if err != nil {
+		return sdk.Coins{}, err
+	}
+
+	coins, err := k.GetClaimable(ctx, addr)
+	if err != nil {
+		return coins, err
+	}
+	percentage := k.GetClaimablePercentageByActivity(ctx, address)
+	withdrawable := sdk.Coins{}
+	for _, coin := range coins {
+		amount := coin.Amount.ToDec().Mul(percentage).RoundInt()
+		if amount.IsPositive() {
+			withdrawable = withdrawable.Add(sdk.NewCoin(coin.Denom, amount))
+		}
+	}
+	return withdrawable, nil
+}
+
+// ClaimCoins remove claimable amount entry and transfer it to user's account
+func (k Keeper) ClaimCoins(ctx sdk.Context, addr string) (sdk.Coins, error) {
```

Comment:
My understanding of the `proto` is that funds are stored by claimable category per user. If thats right, then should we be passing in the activity enum type here, and then have `GetWithdrawableByActivity` take that as in input?

Then `GetWithdrawableByActivity` would read the amount from that directly.

Reply:
Is my understanding of the proto file correct, and if so, does it only apply to the genesis state, not whats stored in state during chain execution?

Reply:
It's predetermined as 25% per action. No need. to get from chain execution - will simplify.